internal/template/utilization_functions.go

```diff
@@ -75,67 +74,95 @@ scalar(
   * 100, 0.0001)
 )`
 
-	return renderUtilization(memoryUtilizationQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelSelectors, memoryUtilizationQueryTemplate)
 }
 
-func cpuRequests(data MetricData, labelArgs ...string) (string, error) {
+func cpuRequests(data MetricData, labelSelectors ...string) (string, error) {
 	cpuResourcesQueryTemplate := `
 scalar(
   sum(
     avg_over_time(kube_pod_container_resource_requests_cpu_cores[{{ .Range }}])
     *
     on (pod) group_left
-    max_over_time(kube_pod_labels{{ .Labels }}[{{ .Range }}])
+    max_over_time(kube_pod_labels{{ .MetricSelector }}[{{ .Range }}])
   )
 )`
 
-	return renderUtilization(cpuResourcesQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelSelectors, cpuResourcesQueryTemplate)
 }
 
-func memoryRequests(data MetricData, labelArgs ...string) (string, error) {
+func memoryRequests(data MetricData, labelSelectors ...string) (string, error) {
 	memoryResourcesQueryTemplate := `
 scalar(
   sum(
     avg_over_time(kube_pod_container_resource_requests_memory_bytes[{{ .Range }}])
     *
     on (pod) group_left
-    max_over_time(kube_pod_labels{{ .Labels }}[{{ .Range }}])
+    max_over_time(kube_pod_labels{{ .MetricSelector }}[{{ .Range }}])
   )
 )`
 
-	return renderUtilization(memoryResourcesQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelSelectors, memoryResourcesQueryTemplate)
 }
 
-func renderUtilization(query string, data MetricData, labelArgs ...string) (string, error) {
-	tmpl := template.Must(template.New("query").Parse(query))
-
-	// Always include the trial namespace
-	labels := []string{fmt.Sprintf("namespace=\"%s\"", data.Trial.Namespace)}
-
-	for _, label := range strings.Split(strings.Join(labelArgs, ","), ",") {
-		if label == "" {
-			continue
-		}
+func renderUtilization(metricData MetricData, labelSelectors []string, query string) (string, error) {
+	// We are accepting Kubernetes label selectors and using them to generate a PromQL metric selector
+	sel, err := labels.Parse(strings.Join(labelSelectors, ","))
+	if err != nil {
+		return "", err
+	}
 
-		kvpair := strings.Split(label, "=")
-		if len(kvpair) != 2 {
-			return "", fmt.Errorf("invalid label for utilization query, expected key=value, got: %s", label)
+	// Always include the trial namespace first
+	requirements, _ := sel.Requirements()
+	labelMatchers := make([]string, 0, 1+len(requirements))
+	labelMatchers = append(labelMatchers, fmt.Sprintf("namespace=%q", metricData.Trial.Namespace))
+	for _, req := range requirements {
```

Comment:
Should we filter out namespace? Kind of a weird/special case since we add it in, but if someone were to specify it in a query we'd try to wrap it with label_.

Reply:
I think `label_namespace` should be the behavior here, we are only trying to let the user query Prometheus metrics based on Kubernetes labels. I think we should stick to generating the `label_` only queries and if people need to query on other Prometheus metric labels, they can do that using their own PromQL.